Adding tests for `prepare.write`

doowb opened the pull request:
This PR adds the tests for `prepare.write` based on previous `vfs.dest` tests from `vinyl-fs`.
Also, this passes the `opt` object into `through.obj`.

doowb committed:
add tests

doowb committed:
udpate call to `through.obj` to pass options through.

phated commented:
Oh, these seem to be based on the testing methods in vinyl-fs before my refactor was merged in.  There's much better patterns now and I don't think this module needs to test certain things (end event, etc).

doowb commented:
> these seem to be based on the testing methods in vinyl-fs before my refactor

Yes, they are.

> There's much better patterns now and I don't think this module needs to test certain things (end event, etc).

I'll update them to the latest patterns. I think it needs tests for the different options too.

doowb committed:
add test utils from vinyl-fs

doowb committed:
throw error when invalid outputFolder is passed to .write

doowb committed:
remove unused test utils

doowb committed:
deps

doowb committed:
update tests to use latest patterns
include tests for options that can be values or functions

doowb committed:
remove check for `basePath` since base option has been removed

doowb committed:
add test for invalid output folder returned from output folder function

doowb commented:
@phated, I updated using the conventions from vinyl-fs and got 100% test coverage.

I removed the possible error being emitted for `basePath` since the `base` option was removed in a previous commit in vinyl-fs.

phated commented:
`base` and `basePath` are different.  `base` was an option we used to support, but `basePath` is a calculated path resolved from `cwd` to `outFolder`

doowb commented:
> `basePath` is a calculated path resolved from `cwd` to `outFolder`

That's why I removed the check for an empty `basePath` since that's already handled by checking `outFolder`. I can put it back in, but I'm not sure how to make it error.

phated commented:
I like a lot of these tests but there are some that don't belong.  I'll merge and amend with my removals.

doowb commented:
Thanks, let me know if there's anything else I can help with.

phated commented:
Will do!  Right now, I'm trying to figure out the best way to handle `prepare.write` as a stream in symlink, specifically https://github.com/gulpjs/vinyl-fs/commit/aaa66028cb08934cb25c18659656cd2a2caf303c#diff-6c822274d928a75825fb98638c458970R24

doowb commented:
What if `vinyl-prepare` exposed a method that returns a stream (like it's doing now), and a method that does the work on an individual file (like it was done in `vinyl-fs`). Then the method doing the work can be used inside the stream when the stream method is used, and also used in other streams (like the `symlink` stream).

The reason I'm thinking this is because doing `this.history[this.history.length - 2]` might be wrong if the filepath didn't actually change in `prepare.write`. (I think this will only happen when the `outFolder` is the same as the `cwd` and would probably cause `symlink` to fail anyway).

phated commented:
@doowb I was thinking that also but it's not really as elegant as I'd like this module to be.  Maybe there needs to be a `prepare.symlink` method that adds an extra property that we can use (probably `.symlink` because `vinylfs.src` adds that property already).

krnlde commented:
What's the state of this PR? Why is it not merged into master to close the corresponding issue?

phated commented:
@krnlde as the PR says at the top, it's already been merged.  There's no corresponding issue to close...

krnlde commented:
True, thank you. I guess I hit the wrong reference. Sorry for the inconvenience.
